### ISSUE TITLE
How to terminate GLFW?

### ISSUE THREAD

**Issue by fmahnke** [PistonDevelopers/glfw-rs#385]
What is the correct way to terminate the GLFW library? I have a test module that runs two tests, each which indirectly create an instance of the GLFW library.

When I don't explicitly terminate GLFW, the first test succeeds and the second test fails as follows:
```
---- test_entity_graph::test_entity_create_delete stdout ----
        thread 'test_entity_graph::test_entity_create_delete' panicked at 'called `Result::unwrap()` on an `Err` value: AlreadyInitialized',
libcore/result.rs:945:5
```
I don't see a safe public function to terminate GLFW, but when I try to explicitly terminate it after each test with `unsafe { glfwTerminate(); }`, this happens, presumably because glfw-rs expects GLFW to be initialized for its internal `drop` routines which run after the `glfwTerminate` call:
```
---- test_entity_graph::test_entity_create_delete stdout ----
        thread 'test_entity_graph::test_entity_create_delete' panicked at 'GLFW Error: The GLFW library is not initialized', /home/fritz/.car
go/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/lib.rs:392:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::_print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic_fmt
   6: glfw::fail_on_errors
             at /home/fritz/.cargo/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/lib.rs:392
   7: <glfw::Callback<for<'r> fn(glfw::Error, alloc::string::String, &'r UserData), UserData> as glfw::callbacks::error::Object<(glfw::Error,
 alloc::string::String)>>::call
             at /home/fritz/.cargo/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/callbacks.rs:45
   8: glfw::callbacks::error::callback::{{closure}}
             at /home/fritz/.cargo/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/callbacks.rs:67
   9: <std::thread::local::LocalKey<T>>::try_with
             at /build/rust/src/rustc-1.27.0-src/src/libstd/thread/local.rs:294
  10: <std::thread::local::LocalKey<T>>::with
             at /build/rust/src/rustc-1.27.0-src/src/libstd/thread/local.rs:248
  11: glfw::callbacks::error::callback
             at /home/fritz/.cargo/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/callbacks.rs:65
  12: _glfwInputError
             at /home/fritz/.cargo/registry/src/github.com-1ecc6299db9ec823/glfw-sys-3.2.2/src/init.c:111
  13: glfwGetWindowUserPointer
             at /home/fritz/.cargo/registry/src/github.com-1ecc6299db9ec823/glfw-sys-3.2.2/src/window.c:785
  14: <glfw::Window as core::ops::drop::Drop>::drop
             at /home/fritz/.cargo/git/checkouts/glfw-rs-0f5796e43f8a68fa/d7f21e1/src/lib.rs:2178
  15: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  16: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  17: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  18: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  19: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  20: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  21: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
             at /build/rust/src/rustc-1.27.0-src/src/liballoc/rc.rs:841
  22: core::ptr::drop_in_place
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ptr.rs:59
  23: ng::test_entity_graph::test_entity_create_delete
             at src/test_entity_graph.rs:60
  24: ng::__test::TESTS::{{closure}}
             at src/test_entity_graph.rs:26
  25: core::ops::function::FnOnce::call_once
             at /build/rust/src/rustc-1.27.0-src/src/libcore/ops/function.rs:223
  26: <F as alloc::boxed::FnBox<A>>::call_box
  27: __rust_maybe_catch_panic
```

**Comment by fmahnke** [comment]
I'm able to get around my problem either by:

1. Storing the GLFW context in a mutable static variable AND using `--test-threads=1` to stop tests in the same source file from running in parallel OR
2. Re-organizing the test code to only one test per file, which causes rust to run them in different processes.

I will go with option #2 for now, and avoid the global.

**Comment by rugamaga** [comment]
I think this issue is now solved by PR #430 
You can drop `Glfw` marker object to terminate GLFW.
